File: app/src/main/java/com/arkapp/partyplanner/ui/venueList/VenueListAdapter.java
package com.arkapp.partyplanner.ui.venueList;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.databinding.DataBindingUtil;
import androidx.navigation.ActivityKt;
import androidx.navigation.NavController;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;

import com.arkapp.partyplanner.R;
import com.arkapp.partyplanner.data.models.PartyDetails;
import com.arkapp.partyplanner.data.models.Venue;
import com.arkapp.partyplanner.data.repository.PrefRepository;
import com.arkapp.partyplanner.data.room.AppDatabase.Companion;
import com.arkapp.partyplanner.data.room.SummaryDetailsDao;
import com.arkapp.partyplanner.databinding.RvVenueBinding;
import com.arkapp.partyplanner.utils.AppDataKt;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import static com.arkapp.partyplanner.utils.ViewUtilsKt.toast;


public final class VenueListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final Gson gson = new Gson();
    private final Type type = new TypeToken<ArrayList<String>>() {
    }.getType();

    private final Activity context;
    private final List<Venue> venueList;
    private final NavController navController;
    private final PrefRepository prefRepository;

    public VenueListAdapter(@NotNull Activity context, @NotNull List<Venue> venueList, @NotNull NavController navController, @NotNull PrefRepository prefRepository) {
        super();
        this.context = context;
        this.venueList = venueList;
        this.navController = navController;
        this.prefRepository = prefRepository;
    }

    @NotNull
    public ViewHolder onCreateViewHolder(@NotNull ViewGroup parent, int viewType) {
        return (ViewHolder) (new VenueListViewHolder((RvVenueBinding) DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.rv_venue, parent, false)));
    }

    @SuppressLint({"ClickableViewAccessibility", "SetTextI18n"})
    public void onBindViewHolder(@NotNull ViewHolder holder, int position) {
        RvVenueBinding binding = ((VenueListViewHolder) holder).getViewBinding();
        final Venue venueData = venueList.get(position);

        binding.venueName.setText(venueData.getName());
        binding.venueAdd.setText(venueData.getAddress());
        binding.capacity.setText(venueData.getCapacity() + " Guest");
        binding.contact.setText(venueData.getContact());
        binding.price.setText("$" + venueData.getPrice());
        binding.location.setText(venueData.getLocation());

        ArrayList<String> partyTypes = gson.fromJson(venueData.getPartyType(), type);
        StringBuilder partyTypeString = new StringBuilder();

        for (String x : partyTypes) {
            partyTypeString.append(x + ", ");
        }
        binding.suitable.setText(partyTypeString.toString().substring(0, partyTypeString.toString().length() - 2));


        binding.parent.setOnClickListener(view -> {
            PartyDetails details = prefRepository.getCurrentPartyDetails();
            details.setSelectedDestination(venueData);
            prefRepository.setCurrentPartyDetails(details);

            if (prefRepository.getCurrentPartyDetails().getCheckedItemList() != null &&
                    !prefRepository.getCurrentPartyDetails().getCheckedItemList().isEmpty()) {
                toast(context, "Please wait saving data...");
                new UpdateSummaryAsyncTask(context, prefRepository).execute();
            } else
                navController.navigate(R.id.action_venueListFragment_to_specialSelectionFragment);
        });
    }

    public int getItemCount() {
        return this.venueList.size();
    }

    public long getItemId(int position) {
        return (long) (venueList.get(position)).hashCode();
    }

    private static final class UpdateSummaryAsyncTask extends AsyncTask<Void, Void, Void> {
        private final Activity context;
        private final PrefRepository prefRepository;

        public UpdateSummaryAsyncTask(@NotNull Activity context, @NotNull PrefRepository prefRepository) {
            super();
            this.context = context;
            this.prefRepository = prefRepository;
        }

        @Nullable
        @Override
        protected Void doInBackground(@NotNull Void... params) {
            SummaryDetailsDao var10000 = (new Companion()).getDatabase((Context) context).summaryDao();
            var10000.delete(prefRepository.getCurrentUser().getUid());
            var10000.insert(AppDataKt.convertSummary(prefRepository.getCurrentPartyDetails(), prefRepository.getCurrentUser().getUid()));
            return null;
        }

        @SuppressLint({"SetTextI18n"})
        @Override
        protected void onPostExecute(@Nullable Void summaryData) {
            ActivityKt.findNavController(context, R.id.fragment).navigate(R.id.action_venueListFragment_to_finalChecklistFragment);
        }
    }
}

File: app/src/main/java/com/arkapp/partyplanner/ui/venueList/VenueListAdapter.kt
package com.arkapp.partyplanner.ui.venueList

import android.annotation.SuppressLint
import android.app.Activity
import android.os.AsyncTask
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.LifecycleCoroutineScope
import androidx.navigation.NavController
import androidx.navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import com.arkapp.partyplanner.R
import com.arkapp.partyplanner.data.models.Venue
import com.arkapp.partyplanner.data.repository.PrefRepository
import com.arkapp.partyplanner.data.room.AppDatabase
import com.arkapp.partyplanner.utils.convertSummary
import com.arkapp.partyplanner.utils.toast
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * Created by Abdul Rehman on 28-02-2020.
 * Contact email - abdulrehman0796@gmail.com
 */
/**
 * Recycler view adapter of venue list venue selection screen
 * */
class VenueListAdapter(
    private val context: Activity,
    private val venueList: List<Venue>,
    private val navController: NavController,
    private val prefRepository: PrefRepository,
    private val lifecycleScope: LifecycleCoroutineScope
) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val gson = Gson()
    private val type = object : TypeToken<ArrayList<String>>() {}.type!!

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return VenueListViewHolder(
            DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.rv_venue,
                parent,
                false
            )
        )
    }

    @SuppressLint("ClickableViewAccessibility", "SetTextI18n")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val binding = (holder as VenueListViewHolder).viewBinding

        val venueData = venueList[position]
        binding.venueName.text = venueData.name
        binding.venueAdd.text = venueData.address
        binding.capacity.text = "${venueData.capacity} Guest"
        binding.contact.text = venueData.contact
        binding.price.text = "$${venueData.price}"
        binding.location.text = venueData.location

        val partyTypes = gson.fromJson<ArrayList<String>>(venueData.partyType, type)
        var partyTypeString = ""

        for (x in partyTypes) {
            partyTypeString += "$x, "
        }
        binding.suitable.text = partyTypeString.substring(0, partyTypeString.length - 2)


        binding.parent.setOnClickListener {
            val details = prefRepository.currentPartyDetails
            details.selectedDestination = venueData
            prefRepository.currentPartyDetails = details

            if (!prefRepository.currentPartyDetails.checkedItemList.isNullOrEmpty()) {
                context.toast("Please wait saving data...")
                UpdateSummaryAsyncTask(context, prefRepository).execute()
            } else
                navController.navigate(R.id.action_venueListFragment_to_specialSelectionFragment)
        }

    }


    override fun getItemCount() = venueList.size

    override fun getItemId(position: Int): Long {
        return venueList[position].hashCode().toLong()
    }

    //adding the selected venue in the SQL db
    private class UpdateSummaryAsyncTask(private val context: Activity,
                                         private val prefRepository: PrefRepository) : AsyncTask<Void, Void, Void?>() {

        override fun doInBackground(vararg params: Void?): Void? {
            val summaryDao = AppDatabase.Companion().getDatabase(context).summaryDao()
            summaryDao.delete(prefRepository.currentUser?.uid!!)
            summaryDao.insert(convertSummary(prefRepository.currentPartyDetails,
                                             prefRepository.currentUser?.uid!!))
            return null
        }

        @SuppressLint("SetTextI18n")
        override fun onPostExecute(summaryData: Void?) {
            context.findNavController(R.id.fragment)
                .navigate(R.id.action_venueListFragment_to_finalChecklistFragment)
        }
    }
}